### PR TITLE
support init rustwide_builder periodically

### PR DESCRIPTION
**Pull request by linyihai**
This is a initial PR for:  https://github.com/rust-lang/docs.rs/issues/2199 (automatically update build docker image)

I create a `peroid_init` method for `docbuilder::rustwide_builder::RustwideBuilder`,  it will init builder again.

Please let me know if this is feasible or if there are other ways of working you prefer.

### CODE REVIEW

File: src/docbuilder/rustwide_builder.rs
Comment (syphar) [inline src/docbuilder/rustwide_builder.rs:1339]: the test environment has `override_config`, which would make the config matching the sleep below more explicit.

File: src/docbuilder/rustwide_builder.rs
Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:112]: I'm aware that most of the content of `RustwideBuilder::init` is building the workspace :)

File: src/docbuilder/rustwide_builder.rs
Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:1339]: This would only test if the build just works, without checking if the workspace was reinitialized. 

Let's finish the refactor above, then we can think of a way to actually test the behaviour.

File: src/config.rs
Comment (syphar) [reply src/config.rs:97]:
```suggestion
    pub(crate) build_workspace_reinitialization_interval: Duration,
```
also we should rename this to explain better what it does, something like this perhaps?

File: src/utils/queue_builder.rs
Comment (syphar) [inline src/utils/queue_builder.rs:22]: we should decide what to do when the re-init fails with any error. 

( whereever the reinit will end up, here or inside `RustwideBuilder`). 

Either we just long an error and continue building crates, though I lean towards locking the queue as we do when the toolchain update fails. 

In any case, just returning with `?` is not good in this place.

File: src/config.rs
Comment (linyihai) [reply src/config.rs:97]: very reasonable 😄

File: src/docbuilder/rustwide_builder.rs
Comment (linyihai) [reply src/docbuilder/rustwide_builder.rs:1339]: I agree with that make the testcase more rubust in the feature.

File: src/utils/queue_builder.rs
Comment (linyihai) [reply src/utils/queue_builder.rs]: i split the code to `build_workspace` which build workspace, and then i rename the `peroid_init` to `reinitialize_workspace_if_interval_passed`, as the name shown, the new function will reinitialize workspace if the interval passed.

**Review by syphar — APPROVED** [review]
From a code & test perspective I'm fine with this change now, and very much looking forward to it. 

In my mind I probably would set the refresh frequency to something shorter than a day, also to directly see potential errors. 

Before I merge this I would only like to run some local tests, though I guess these will be fine.

**Comment by syphar** [comment]
@linyihai recent merges lead to conflicts, could you rebase your code onto master?

**Comment by syphar** [comment]
test was successful, after rebase I'll merge

**Comment by linyihai** [comment]
> @linyihai recent merges lead to conflicts, could you rebase your code onto master?

Done

**Pull request merged**
